### PR TITLE
Entities Audit Log

### PR DESCRIPTION
**Pull request by romanetar**
https://tipit.avaza.com/project/view#!tab=task-pane&groupby=MyTaskDueDate&view=vertical&task=3084385&fileview=grid

Signed-off-by: romanetar <roman_ag@hotmail.com>

### CODE REVIEW

File: app/Models/Foundation/Main/Audit/AuditLog.php [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+ * Class AuditLog
+ * @package models\main
+ */
+abstract class AuditLog extends SilverstripeBaseModel
+{
+    /**
+     * @ORM\Column(name="User", type="string")
+     * @var string|null
+     */
+    private $user;
```
Comment (smarcet) [inline app/Models/Foundation/Main/Audit/AuditLog.php:35]: user should be a Member not a string should point to member table ( FK)

File: app/Models/Foundation/Summit/SelectionPlan.php [hunk 2]
```diff
@@ -48,6 +48,7 @@
  *     )
  * })
  * @ORM\Table(name="SelectionPlan")
+ * @ORM\HasLifecycleCallbacks
```
Comment (smarcet) [inline app/Models/Foundation/Summit/SelectionPlan.php:51]: do we need this one

File: database/migrations/model/Version20221124175204.php [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        $builder = new Builder($schema);
+        if (!$builder->hasTable("AuditLog")) {
+            $builder->create("AuditLog", function (Table $table) {
+                $table->integer("ID", true, false);
+                $table->primary("ID");
+                $table->string('ClassName', 255)->setDefault("AuditLog");
+                $table->index("ClassName", "ClassName");
+                $table->timestamp('Created');
+                $table->timestamp('LastEdited');
+                $table->string("User", 255)->setNotnull(false);
```
Comment (smarcet) [inline database/migrations/model/Version20221124175204.php:26]: this should be a FK to Member table

File: app/Services/Utils/DoctrineTransactionService.php [hunk 4]
```diff
@@ -59,6 +62,11 @@ public function transaction(Closure $callback, int $isolationLevel = Transaction
         while (!$done and $retry < self::MaxRetries) {
             try {
                 $em  = Registry::getManager($this->manager_name);
+                $event_manager = $em->getEventManager();
+
+                if (!$event_manager->hasListeners(Events::onFlush))
```
Comment (smarcet) [inline app/Services/Utils/DoctrineTransactionService.php:67]: i do think that this could be done on doctrine config not here

**Review by smarcet — CHANGES_REQUESTED** [review]
@romanetar please review comments

File: app/Audit/Loggers/SummitEventAuditLogger.php [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+        $user_id = $resource_server_ctx->getCurrentUserId();
+
+        $entry = new SummitEventAuditLog(
+            $user_id,
+            $description,
+            $entity->getSummit(),
+            $entity
+        );
+        $entity_manager->persist($entry);
+        $entity_manager->getUnitOfWork()->computeChangeSet($entity_manager->getClassMetadata(get_class($entry)), $entry);
```
[the remark this answers is not in the document]
Comment (romanetar) [reply app/Audit/Loggers/SummitEventAuditLogger.php:48]: it's the way to tell the current entity manager that we need to persist a new entity (in this case, AuditLog). We can't call flush again at this point

File: app/Models/Foundation/Summit/SelectionPlan.php (hunk 2 above)
Comment (romanetar) [reply app/Models/Foundation/Summit/SelectionPlan.php:51]: Not really, I gonna remove this one.

File: app/Models/Foundation/Main/Audit/AuditLog.php (hunk 1 above)
Comment (romanetar) [reply app/Models/Foundation/Main/Audit/AuditLog.php:35]: done

File: app/Models/Foundation/Summit/SelectionPlan.php (hunk 2 above)
Comment (romanetar) [reply app/Models/Foundation/Summit/SelectionPlan.php:51]: done

File: database/migrations/model/Version20221124175204.php (hunk 3 above)
Comment (romanetar) [reply database/migrations/model/Version20221124175204.php:26]: done

File: app/Services/Utils/DoctrineTransactionService.php (hunk 4 above)
Comment (romanetar) [reply app/Services/Utils/DoctrineTransactionService.php:67]: done

**Comment by romanetar** [comment]
@smarcet all resolved here https://github.com/OpenStackweb/summit-api/commit/c5b66dfdc62cb338697c8679c918952623293a4d

**Review by smarcet — CHANGES_REQUESTED** [review]
@romanetar please review comments

File: app/Audit/Loggers/SummitEventAuditLogger.php (hunk 5 above)
[the remark this answers is not in the document]
Comment (romanetar) [reply app/Audit/Loggers/SummitEventAuditLogger.php:56]: done

File: app/Repositories/Main/DoctrineAuditLogRepository.php [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+ * @package repositories\main
+ */
+final class DoctrineAuditLogRepository
+    extends SilverStripeDoctrineRepository
+    implements IAuditLogRepository
+{
+    protected function getFilterMappings(): array
+    {
+        return [
+            'user' => 'e.user:json_string',
```
[the remark this answers is not in the document]
Comment (romanetar) [reply app/Repositories/Main/DoctrineAuditLogRepository.php]: done

**Comment by romanetar** [comment]
@smarcet all review issues addressed here: https://github.com/OpenStackweb/summit-api/commit/34e01434ff15805c4e71dd706134e49f6bf04b03

File: app/ModelSerializers/Audit/AuditLogSerializer.php [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+     * @return array
+     */
+    public function serialize($expand = null, array $fields = [], array $relations = [], array $params = [])
+    {
+        $audit_log = $this->object;
+        if (!$audit_log instanceof AuditLog) return [];
+
+        $values = parent::serialize($expand, $fields, $relations, $params);
+
+        if (!empty($expand)) {
```
Comment (smarcet) [inline app/ModelSerializers/Audit/AuditLogSerializer.php:53]: here u could use One2ManyExpandSerializer like here https://github.com/OpenStackweb/summit-api/blob/4db85f6854f896cd4b739ccefc482ad0296480a1/app/ModelSerializers/Elections/NominationSerializer.php#L36

File: app/ModelSerializers/Audit/SummitAuditLogSerializer.php [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+        if (!$summit_audit_log instanceof SummitAuditLog) return [];
+
+        $values = parent::serialize($expand, $fields, $relations, $params);
+
+        if (!empty($expand)) {
+            foreach (explode(',', $expand) as $relation) {
+                $relation = trim($relation);
+                if ($relation == 'summit') {
+                    unset($values['summit_id']);
+                    $values['summit'] = SerializerRegistry::getInstance()->getSerializer($summit_audit_log->getSummit())->serialize(AbstractSerializer::filterExpandByPrefix($expand, $relation));
```
Comment (smarcet) [inline app/ModelSerializers/Audit/SummitAuditLogSerializer.php:48]: same here https://github.com/OpenStackweb/summit-api/blob/4db85f6854f896cd4b739ccefc482ad0296480a1/app/ModelSerializers/Elections/NominationSerializer.php#L36

File: app/ModelSerializers/Audit/SummitEventAuditLogSerializer.php [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+    public function serialize($expand = null, array $fields = [], array $relations = [], array $params = [])
+    {
+        $summit_event_audit_log = $this->object;
+        if (!$summit_event_audit_log instanceof SummitEventAuditLog) return [];
+
+        $values = parent::serialize($expand, $fields, $relations, $params);
+
+        if (!empty($expand)) {
+            foreach (explode(',', $expand) as $relation) {
+                $relation = trim($relation);
```
Comment (smarcet) [inline app/ModelSerializers/Audit/SummitEventAuditLogSerializer.php:45]: same here https://github.com/OpenStackweb/summit-api/blob/4db85f6854f896cd4b739ccefc482ad0296480a1/app/ModelSerializers/Elections/NominationSerializer.php#L36

File: app/Models/Foundation/Main/Audit/SummitEventAuditLog.php [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+     * @var SummitEvent
+     */
+    private $event;
+
+    public function getClassName(): string
+    {
+        return self::ClassName;
+    }
+
+    public function getEventId(): ?int
```
Comment (smarcet) [inline app/Models/Foundation/Main/Audit/SummitEventAuditLog.php:40]: methods like getXYZId and hasXYZ could be implemented using trait One2ManyPropertyTrait;
like here

File: app/ModelSerializers/Audit/AuditLogSerializer.php (hunk 7 above)
Comment (romanetar) [reply app/ModelSerializers/Audit/AuditLogSerializer.php]: done

File: app/ModelSerializers/Audit/SummitAuditLogSerializer.php (hunk 8 above)
Comment (romanetar) [reply app/ModelSerializers/Audit/SummitAuditLogSerializer.php]: done

File: app/Models/Foundation/Main/Audit/SummitEventAuditLog.php (hunk 10 above)
Comment (romanetar) [reply app/Models/Foundation/Main/Audit/SummitEventAuditLog.php]: done

File: database/migrations/model/Version20221124175204.php [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+                $table->integer("ID", true, false);
+                $table->primary("ID");
+                $table->string('ClassName', 255)->setDefault("AuditLog");
+                $table->index("ClassName", "ClassName");
+                $table->timestamp('Created');
+                $table->timestamp('LastEdited');
+                $table->text("Action")->setNotnull(true);
+                $table->integer("UserID", false, false)->setNotnull(false)->setDefault('NULL');
+                $table->index("UserID", "UserID");
+                $table->foreign("Member", "UserID", "ID", ["onDelete" => "CASCADE"]);
```
[the remark this answers is not in the document]
Comment (romanetar) [reply database/migrations/model/Version20221124175204.php]: done

File: app/Models/Foundation/Main/Audit/AuditLog.php [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+
+    /**
+     * @ORM\Column(name="Action", type="string")
+     * @var string
+     */
+    private $action;
+
+    public function getUserId(): ?int
```
[the remark this answers is not in the document]
Comment (romanetar) [reply app/Models/Foundation/Main/Audit/AuditLog.php]: done

**Comment by romanetar** [comment]
@smarcet all review issues addressed here: https://github.com/OpenStackweb/summit-api/commit/2384129ee1dfcd0cbef88d7da55431afbd5495dd

**Review by smarcet — APPROVED** [review]
LGTM

**Pull request merged**